### ISSUE TITLE
OpenID or SAML authentication

### ISSUE THREAD

**Issue by sachz19** [opnsense/core#6110]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [X] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [X] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Describe the solution you like**

I would like Oauth, OpenID or SAML to be added to the authentication servers.
We are running our services completely through our central authentication provider, but LDAP is currently not an option.
Adding multiple ways of authenticating might also be interesting for captive portal authentication, VPN authentication and other possible implementations that might come into the future (statistics plugins , ...)

**Describe alternatives you considered**

considered using a different setup which would makee ldap available, but the current setup is desired for our current posture.

**Additional context**

Adding those authentication protocols might be interesting.
Expanding the central users to the different plugins might be an interesting next step.
This way you could use external OpenID provider to allow access to a site hosted through haproxy / Nginx reverse proxy
Web Proxy authentication, ....

**Comment by AdSchellevis** [comment]
Practically I don't think this will work as a generic concept in our solution, but always open for suggestions on how this technically should work while also offering the same experience for other consumers like OpenVPN and IPsec. Similar discussions have been taken place in the past:

https://github.com/opnsense/core/issues/1855
https://github.com/opnsense/core/issues/3294

**Comment by sachz19** [comment]
I do understand the hesitation, definitely when talking about SAML.
Being able to authenticate towards multiple sources will just provide a broader connection experience.
LDAP is no longer the Go-To directory server for more and more businesses, azure Ad is a nice example of the move for the classic smb environments.
Allowing multiple backends (like you can do now with local authentication and LDAP) with an extra source can also fullfill some requirements when companies are going towards certain ISO certified environment.

In my opinion, the first step would be to provide an extra authentication mechanism for OpenID, just like ldap is currently implemented.
then look into connecting the authentication to openvpn configuration just like the ldap connection is done.

For me personally the most important parts are authentication towards the opnsense gui and use with the ssl vpn (openvpn)
(For ssh i'm not sure if it is doable with OpenID)

​​​​​If you look at other opensource solutions, most of them have started work to implement OpenID.
By the use of the external authentication server you also allow more granular control over conditional access, Security keys, ....

​​​​​
​Since opnsense is a security product, it might be interesting to start implementing ways of complying with multiple types of environments.
Currently we for instance have no LDAP, and since thats the only real directory service you can externally connect to, the options seem limited.



On Wednesday, 02 November, 2022 13:20 CET, Ad Schellevis ***@***.***> wrote:
 
 
Practically I don't think this will work as a generic concept in our solution, but always open for suggestions on how this technically should work while also offering the same experience for other consumers like OpenVPN and IPsec. Similar discussions have been taken place in the past:
#1855
#3294

**Comment by AdSchellevis** [comment]
Try to think this through and try to describe how the workflow should operate from the viewpoint of an application (OpenVPN for example) which has no web form for validation. 

> In my opinion, the first step would be to provide an extra authentication mechanism for OpenID, just like ldap is currently implemented.

The advised flows for OpenID, Oauth and SAML in my understanding all operate within the concept that they operate with a "letter of understanding" from the authentication provider, which means the flow doesn't even remotely look like a user/password flow which we implement (https://docs.opnsense.org/development/components/authentication.html).  

In theory one could  implement the resource owner password flow (https://auth0.com/docs/get-started/authentication-and-authorization-flow/resource-owner-password-flow), although in most cases that doesn't really add anything as these providers also support LDAP and/or RADIUS for these purposes.

Open for suggestions, but as it stands, I don't see a valid path implementing OAuth/OpenID/SAML in a way that components within OPNsense are able to use the connector (in a similar way as LDAP)

**Comment by FirefoxNL** [comment]
OPNsense will support OpenID Connect starting from Business edition 25.10. [Source](https://forum.opnsense.org/index.php?topic=41661.msg240620#msg240620). Not planned for community edition so far apparently.

I stumbled upon the reply from Franco and a simple Google search still brings this issue up as one of the first results so I figured it might be worth it to post this here for anyone searching.
